zerver/webhooks/semaphore2/view.py

```diff
@@ -0,0 +1,180 @@
+# Webhooks for external integrations.
+from typing import Any, Dict, List
+
+from django.http import HttpRequest, HttpResponse
+
+from zerver.decorator import api_key_only_webhook_view
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.webhooks.common import check_send_webhook_message
+from zerver.models import UserProfile
+from urllib.parse import urlparse
+
+# There seems to be a possibility of only 2 types of events : Push or a
+# Pull request, which can trigger a pipeline.
+
+# As per the Semaphore2.0 documentation, notifications would be sent only when
+# a pipeline is triggered.
+
+PUSH_TEMPLATE = """
+**{email}** pushed to branch **{branch_name}** of \
+[{organization_name}/{project_name}]({repository_url}):
+* **Commit**: [(#{commit_hash}) {commit_message}]({commit_url})
+* **Pipeline**:
+  - Name: {pipeline_name}
+  - Result: {pipeline_result}
+* **Blocks**:
+{blocks}
+""".strip()
+
+PULL_REQUEST_TEMPLATE = """
+**{email}** created a pull request to branch **{branch_name}** of \
+[{organization_name}/{project_name}]({repository_url}):
+* **Pull Request**: [{pull_request_title}]({pull_request_url})
+* **Pipeline**:
+  - Name: {pipeline_name}
+  - Result: {pipeline_result}
+* **Blocks**:
+{blocks}
+""".strip()
+
+DEFAULT_TEMPLATE = """
+An activity occured on branch **{branch_name}** of \
+[{organization_name}/{project_name}]({repository_url}):
+* **Author**: {email}
+* **Pipeline**:
+  - Name: {pipeline_name}
+  - Result: {pipeline_result}
+* **Blocks**:
+{blocks}
+""".strip()
+
+TOPIC_TEMPLATE = "{project}/{branch}"
+
+@api_key_only_webhook_view('Semaphore')
+@has_request_variables
+def api_semaphore2_webhook(request: HttpRequest, user_profile: UserProfile,
+                           payload: Dict[str, Any]=REQ(argument_type='body')) -> HttpResponse:
+
+    organization_name = payload["organization"]["name"]
+    project_name = payload["project"]["name"]
+    repository_url = payload["repository"]["url"]
+    pipeline_name = payload["pipeline"]["name"]
+    pipeline_result = payload["pipeline"]["result"]
+    author_email = payload["revision"]["sender"]["email"]
+    branch_name = payload["revision"]["branch"]["name"]
+    blocks = get_blocks_data(payload["blocks"])
+
+    if payload["revision"]["pull_request"] is None:
```

Comment:
I looked at the semaphore docs again, and they say

> In this example `revision.pull_request` and `revision.tag` are null because payload is related to the pipeline run started from a push to the branch. Information about this is kept in `revision.reference_type`.

The documentation is not clear on what the possible values for `reference_type` are, but I suspect it would be different for a `pull_request`. You might have to play with Semaphore to figure this out. 

Also, it sounds like it is possible to have events triggered when a `tag` is pushed, because there's a `revision.tag` field they mention? So, this check on the payload may not be sufficient, and we can do better, I think.

Reply:
I think you're right. I just ignored the `revision.tag` field, there must be some significance of it. I'll try to look deeper into the semaphore docs and try to figure out every possible event which can trigger a notification.